src/storage/src/source/reclock.rs

```diff
@@ -8,426 +8,327 @@
-                                .push_back((
-                                    timestamp,
-                                    MzOffset {
-                                        offset: diff
-                                            .try_into()
-                                            .expect("reclock offset diff to be positive"),
-                                    },
-                                ));
-                        }
-                    }
+                    Ok(Ok(Ok(()))) => break,
+                    Ok(Ok(Err(actual_upper))) => return Err(actual_upper),
+                    Ok(Err(invalid_use)) => panic!("{invalid_use:?}"),
+                    Err(_external_err) => continue,
                 }
             }
+            self.sync(&new_upper).await;
+        }
+        Ok(())
+    }
+
+    /// Calling this method promises that all future timestamping requests will be for offsets
+    /// greater or equal to the provided source compaction frontier. This gives permission to the
+    /// reclock operator to compact and consolidate its internal state accordingly.
+    ///
+    /// If the given source since frontier results in a change of the dest since frontier it will
+    /// be returned to the caller who can then use it to downgrade its capability.
+    pub async fn compact(
+        &mut self,
+        source_since: &HashMap<PartitionId, MzOffset>,
+    ) -> Option<AntichainRef<'_, Timestamp>> {
+        // Translate the SourceTime compaction frontier into a DestTime compaction frontier
+        let new_since = source_since
+            .iter()
+            .map(|(pid, part_since)| {
+                self.partition_uppers(pid)
+                    .find(|(_ts, offset)| part_since < offset)
+                    .expect("attempt to compact past upper")
+                    .0
+            })
+            .min()
+            .map(Antichain::from_elem)
+            .unwrap_or_else(|| self.since.clone());
+
+        if PartialOrder::less_than(&self.since, &new_since) {
+            for bindings in self.remap_trace.values_mut() {
+                // Compact the remap trace according to the computed frontier
+                for (timestamp, _) in bindings.iter_mut() {
+                    timestamp.advance_by(new_since.borrow());
+                }
+                // And then consolidate
+                consolidation::consolidate(bindings);
+            }
+            self.since = new_since;
+            self.read_handle.downgrade_since(self.since.clone()).await;
+            Some(self.since.borrow())
+        } else {
+            None
+        }
+    }
+
+    /// Produces a new timestamp suitable for minting bindings. The returned value is guaranteed to
+    /// be beyond the current `upper` frontier and a multiple of `update_interval_ms`.
+    async fn next_mint_timestamp(&mut self) -> Timestamp {
+        loop {
+            let mut new_ts = (self.now)();
+            new_ts -= new_ts % self.update_interval_ms;
```

Comment:
this is `floor`ing the `new_ts` which means we could timestamp something at a time _before_ when it enters the system.  I'm mentioning this as purely a UX problem because it seems we could `ceil` instead of `floor` and we should have the same functionality.  Is there a reason we shouldn't do this?

If the answer is that we need the result of `next_mint_timestamp` to strictly go forward, I'd be worried because `self.now` is not monotonic so that can already happen.  Can we also make a TODO for that? Very okay with me if that's the same as the one I'm suggestion for handling clock skew

Reply:
No it doesn't have to do with monotonicity, ceiling would work just as well. I don't know how much of a UX problem it is but I'll change it to a ceil.

Regarding monotonicity, the code already guards for that by looping until we get a timestamp that is beyond the current upper frontier.